File: backend/routers/license.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2024-2026 courseManage Contributors
"""授权门禁占位模块（本部署已放开全部高级功能）。

背景
----
上游版本通过 RSA 签名的 License Key 对 9 项高级功能做门禁，并会向供应商的
企业微信 webhook / 心跳服务回传学校名称、联系人、机器码等信息。

本仓库为 AGPL-3.0 下的自用分支，已：
  * 删除 ``utils/license.py``（RSA 验签、机器码、心跳、供应商发现）
  * 移除 ``main.py`` 中的 ``premium_feature_guard`` 中间件
  * 移除全部向供应商回传数据的接口

为什么保留本文件而不是一并删除：
  ``_check_premium_feature`` 被 11 个模块（fees / grades / evaluations /
  database / wechat / settings / schedules / smart_command 等）引用。
  保留这一个恒真函数，就能一次性放开所有调用点，无需改动那 11 个文件，
  改动面最小、回归风险最低。

同时保留少量只读接口，让前端「系统授权管理」页面仍能正常渲染。
"""

from typing import List, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from database import get_db
from models import Settings
from routers.auth import get_current_user, User

router = APIRouter(prefix="/license", tags=["系统授权"])

# 9 项原高级功能标识（保留常量，供前端与其他模块引用）
PREMIUM_FEATURES = [
    "grade_trend",
    "fee_management",
    "smart_scheduling",
    "wechat_notify",
    "smart_command",
    "dashboard_view",
    "floating_sphere",
    "database_management",
    "student_evaluation",
]

FEATURE_NAMES = {
    "grade_trend": "学生成绩管理",
    "fee_management": "费用管理",
    "smart_scheduling": "智能算法排课",
    "wechat_notify": "微信通知管理",
    "smart_command": "智能指令管理",
    "dashboard_view": "运营大屏",
    "floating_sphere": "全站快捷按钮",
    "database_management": "数据库管理",
    "student_evaluation": "学生评价管理",
}

_OPEN_NOTICE = "本部署已放开全部高级功能，无需授权。"


def _check_premium_feature(feature_name: str, db: Session) -> bool:
    """恒返回 True —— 门禁已移除。

    签名与上游保持一致（``feature_name`` / ``db``），因此那 11 个调用点
    无需任何修改。参数刻意保留但不使用。
    """
    return True


def _all_features_enabled() -> dict:
    return {feature: True for feature in PREMIUM_FEATURES}


# ---------------------------------------------------------------- 只读兼容接口
class FeatureCheckResponse(BaseModel):
    feature: str
    enabled: bool


class LicenseStatusResponse(BaseModel):
    activated: bool = True
    license_type: str = "perpetual"
    license_type_name: str = "自用版（已放开全部功能）"
    organization_name: str = ""
    features: dict = {}
    expiry_date: Optional[str] = None
    issued_at: Optional[str] = None
    machine_code: str = ""
    trial_available: bool = False
    deactivated_licenses: list = []
    license_key: str = ""
    referral_code: str = ""
    referral_activated: bool = False
    referral_threshold: float = 0
    discount_percent: float = 0
    rebate_percent: float = 0
    total_spending: float = 0
    site_name: str = ""
    contact_person: str = ""
    contact_phone: str = ""
    contact_email: str = ""
    contact_wechat: str = ""
    notice: str = _OPEN_NOTICE


@router.get("/status", response_model=LicenseStatusResponse)
def get_license_status(db: Session = Depends(get_db)):
    """返回「已激活、全部功能可用」的固定状态。

    仍从 settings 读取学校与联系人信息，让前端页面展示保持正常。
    """
    settings = db.query(Settings).first()

    def field(name: str) -> str:
        if not settings:
            return ""
        return getattr(settings, name, "") or ""

    return LicenseStatusResponse(
        features=_all_features_enabled(),
        organization_name=field("site_name"),
        site_name=field("site_name"),
        contact_person=field("contact_person"),
        contact_phone=field("contact_phone"),
        contact_email=field("contact_email"),
        contact_wechat=field("contact_wechat"),
    )


@router.get("/features", response_model=List[FeatureCheckResponse])
def list_features(current_user: User = Depends(get_current_user)):
    """列出全部功能，均为启用状态。"""
    return [FeatureCheckResponse(feature=f, enabled=True) for f in PREMIUM_FEATURES]


@router.get("/check/{feature_name}", response_model=FeatureCheckResponse)
def check_feature(feature_name: str, current_user: User = Depends(get_current_user)):
    """单个功能检查，恒为启用。"""
    return FeatureCheckResponse(feature=feature_name, enabled=True)


@router.get("/machine-code")
def get_machine_code(current_user: User = Depends(get_current_user)):
    """机器码机制已随授权校验一并移除。

    保留接口只为兼容前端调用；不再读写 ``/app/backups/.machine_id``，
    因此删除数据卷也不会再导致「授权失效」。
    """
    return {"machine_code": "", "notice": "机器码机制已移除（门禁已放开）。"}


# ---------------------------------------------------------------- 写操作占位
class LicenseActionResponse(BaseModel):
    success: bool
    message: str


@router.post("/activate", response_model=LicenseActionResponse)
def activate_license(current_user: User = Depends(get_current_user)):
    """无需激活。保留接口避免前端按钮 404。"""
    return LicenseActionResponse(success=True, message=_OPEN_NOTICE)


@router.post("/deactivate", response_model=LicenseActionResponse)
def deactivate_license(current_user: User = Depends(get_current_user)):
    """不提供停用能力：门禁已移除，停用没有意义。"""
    return LicenseActionResponse(
        success=False, message="门禁已移除，无法停用；如需恢复门禁请还原 routers/license.py。"
    )


@router.post("/deactivate-feature/{feature_name}", response_model=LicenseActionResponse)
def deactivate_feature(feature_name: str, current_user: User = Depends(get_current_user)):
    """同上，单功能停用同样不再支持。"""
    return LicenseActionResponse(
        success=False, message="门禁已移除，无法停用单个功能。"
    )


# ---------------------------------------------------------------- 已禁用的对外回传
# 上游这几个接口会把学校名称、联系人、电话、邮箱、机器码等发送到供应商的
# 企业微信 webhook 与邮箱。自用部署没有这个需要，故只保留空实现，
# 既不发送任何数据，也不会让前端按钮变成 404。
_NO_UPSTREAM = "本部署已移除向供应商回传数据的通道，未发送任何信息。"


@router.post("/apply", response_model=LicenseActionResponse)
def apply_license(payload: Optional[dict] = None, current_user: User = Depends(get_current_user)):
    """原「申请授权」：不再向供应商发送学校与联系人信息。"""
    return LicenseActionResponse(success=False, message=_NO_UPSTREAM)


@router.post("/preview-addon", response_model=LicenseActionResponse)
def preview_addon_license(payload: Optional[dict] = None, current_user: User = Depends(get_current_user)):
    """原「预览增购」：授权体系已移除。"""
    return LicenseActionResponse(success=False, message=_OPEN_NOTICE)


@router.post("/notify-supplier-view", response_model=LicenseActionResponse)
def notify_supplier_view(payload: Optional[dict] = None):
    """原「通知供应商查看」：已禁用。"""
    return LicenseActionResponse(success=False, message=_NO_UPSTREAM)


@router.post("/request-replace", response_model=LicenseActionResponse)
def request_replace_license(payload: Optional[dict] = None):
    """原「申请换机」：机器码机制已移除，不再需要换机。"""
    return LicenseActionResponse(success=False, message=_NO_UPSTREAM)


@router.post("/feedback", response_model=LicenseActionResponse)
def submit_feedback(payload: Optional[dict] = None):
    """原「功能/系统反馈」：不再回传到供应商 webhook。

    如需保留反馈功能，可在此改为只发送到本校自己配置的 SMTP 邮箱。
    """
    return LicenseActionResponse(success=False, message=_NO_UPSTREAM)

File: frontend/src/utils/license.js
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2024-2026 courseManage Contributors
/**
 * License 状态管理
 * 前端全局单例
 */
import { reactive } from 'vue'
import api from './api'
import i18n from '@/locales'

const { t } = i18n.global

export const FEATURES = {
  GRADE_TREND: 'grade_trend',
  FEE_MANAGEMENT: 'fee_management',
  SMART_SCHEDULING: 'smart_scheduling',
  WECHAT_NOTIFY: 'wechat_notify',
  SMART_COMMAND: 'smart_command',
  DASHBOARD_VIEW: 'dashboard_view',
  FLOATING_SPHERE: 'floating_sphere',
  DATABASE_MANAGEMENT: 'database_management',
  STUDENT_EVALUATION: 'student_evaluation',
}

export const FEATURE_NAMES = {
  [FEATURES.GRADE_TREND]: 'license.featureGradeTrend',
  [FEATURES.FEE_MANAGEMENT]: 'license.featureFeeManagement',
  [FEATURES.SMART_SCHEDULING]: 'license.featureSmartScheduling',
  [FEATURES.WECHAT_NOTIFY]: 'license.featureWechatNotify',
  [FEATURES.SMART_COMMAND]: 'license.featureSmartCommand',
  [FEATURES.DASHBOARD_VIEW]: 'license.featureDashboardView',
  [FEATURES.FLOATING_SPHERE]: 'license.featureFloatingSphere',
  [FEATURES.DATABASE_MANAGEMENT]: 'license.featureDatabaseManagement',
  [FEATURES.STUDENT_EVALUATION]: 'license.featureStudentEvaluation',
}

export const FEATURE_DESCRIPTIONS = {
  [FEATURES.GRADE_TREND]: {
    highlights: 'license.featureGradeTrendHighlights',
    details: 'license.featureGradeTrendDetails',
  },
  [FEATURES.FEE_MANAGEMENT]: {
    highlights: 'license.featureFeeManagementHighlights',
    details: 'license.featureFeeManagementDetails',
  },
  [FEATURES.SMART_SCHEDULING]: {
    highlights: 'license.featureSmartSchedulingHighlights',
    details: 'license.featureSmartSchedulingDetails',
  },
  [FEATURES.WECHAT_NOTIFY]: {
    highlights: 'license.featureWechatNotifyHighlights',
    details: 'license.featureWechatNotifyDetails',
  },
  [FEATURES.SMART_COMMAND]: {
    highlights: 'license.featureSmartCommandHighlights',
    details: 'license.featureSmartCommandDetails',
  },
  [FEATURES.DASHBOARD_VIEW]: {
    highlights: 'license.featureDashboardViewHighlights',
    details: 'license.featureDashboardViewDetails',
  },
  [FEATURES.FLOATING_SPHERE]: {
    highlights: 'license.featureFloatingSphereHighlights',
    details: 'license.featureFloatingSphereDetails',
  },
  [FEATURES.DATABASE_MANAGEMENT]: {
    highlights: 'license.featureDatabaseManagementHighlights',
    details: 'license.featureDatabaseManagementDetails',
  },
  [FEATURES.STUDENT_EVALUATION]: {
    highlights: 'license.featureStudentEvaluationHighlights',
    details: 'license.featureStudentEvaluationDetails',
  },
}

export const LICENSE_TYPES = {
  trialA:     { name: 'license.trialLicense', days: 3 },
  trialB:     { name: 'license.trialLicense', days: 7 },
  monthly:    { name: 'license.monthlyLicense', days: 30 },
  quarterly:  { name: 'license.quarterlyLicense', days: 90 },
  semiannual: { name: 'license.semiannualLicense', days: 180 },
  annual:     { name: 'license.annualLicense', days: 365 },
  perpetual:  { name: 'license.perpetualLicense', days: null },
}

export const FEATURE_PRICES = {
  [FEATURES.FLOATING_SPHERE]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.SMART_COMMAND]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.DASHBOARD_VIEW]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.WECHAT_NOTIFY]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.GRADE_TREND]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.FEE_MANAGEMENT]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.SMART_SCHEDULING]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.DATABASE_MANAGEMENT]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
  [FEATURES.STUDENT_EVALUATION]: { trialA: 1.99, trialB: 2.99, monthly: 6.99, quarterly: 17.99, semiannual: 29.99, annual: 49.99, perpetual: 196.99 },
}

export function calcTotalPrice(selectedFeatures, licenseType) {
  if (!selectedFeatures.length || !licenseType) return 0
  let total = 0
  for (const feat of selectedFeatures) {
    const priceTable = FEATURE_PRICES[feat]
    if (priceTable && priceTable[licenseType] !== undefined) {
      total += priceTable[licenseType]
    }
  }
  return total
}

// 本部署已放开全部功能门禁，但有些模块并不适用于中小学教务场景，
// 在这里集中关闭。关闭后：菜单、页面按钮、运营大屏图表、快捷球入口
// 以及路由守卫都会自动隐藏/拦截 —— 复用的是各处已有的 hasFeature 判断，
// 不需要逐个文件去删模板，改动面最小。
//
// fee_management：课时费/缴费/退费/催缴。义务教育阶段不向学生按课时收费，
//                 后端 /api/fees/* 接口与相关报表已一并移除。
export const DISABLED_FEATURES = [
  FEATURES.FEE_MANAGEMENT,
]

// 本部署已放开全部高级功能：初始状态即为「已激活 + 全部功能可用」。
// 这样即使 /license/status 请求失败或后端未就绪，路由守卫与界面也不会误锁功能。
const ALL_FEATURES_ENABLED = Object.values(FEATURES)
  .filter((feature) => !DISABLED_FEATURES.includes(feature))
  .reduce((acc, feature) => { acc[feature] = true; return acc }, {})

export const licenseState = reactive({
  loaded: true,
  activated: true,
  licenseType: 'perpetual',
  licenseTypeName: '',
  organizationName: '',
  features: { ...ALL_FEATURES_ENABLED },
  expiryDate: null,
  issuedAt: null,
  machineCode: '',
  trialAvailable: false,
  deactivatedLicenses: [],
  licenseKey: '',
  referralCode: '',
  referralActivated: false,
  referralThreshold: 0,
  discountPercent: 0,
  rebatePercent: 0,
  totalSpending: 0,
  siteName: '',
  contactPerson: '',
  contactPhone: '',
  contactEmail: '',
  contactWechat: '',
})

export async function loadLicenseStatus() {
  try {
    const response = await api.get('/license/status')
    const d = response.data
    // 只同步机构/联系人等展示信息；activated 与 features 恒为全开，
    // 不受后端返回值影响（router/index.js 的守卫直接读这两个字段）。
    Object.assign(licenseState, {
      loaded: true,
      activated: true,
      licenseType: d.license_type || 'perpetual',
      licenseTypeName: d.license_type_name || '',
      organizationName: d.organization_name || '',
      features: { ...ALL_FEATURES_ENABLED },
      expiryDate: null,
      issuedAt: d.issued_at || null,
      machineCode: d.machine_code || '',
      trialAvailable: false,
      deactivatedLicenses: [],
      licenseKey: '',
      referralCode: d.referral_code || '',
      referralActivated: d.referral_activated || false,
      referralThreshold: d.referral_threshold || 0,
      discountPercent: d.discount_percent != null ? d.discount_percent : 0,
      rebatePercent: d.rebate_percent != null ? d.rebate_percent : 0,
      totalSpending: d.total_spending || 0,
      siteName: d.site_name || '',
      contactPerson: d.contact_person || '',
      contactPhone: d.contact_phone || '',
      contactEmail: d.contact_email || '',
      contactWechat: d.contact_wechat || '',
    })
  } catch (error) {
    // 后端不可达时也保持全开，避免误锁
    licenseState.loaded = true
    licenseState.activated = true
    licenseState.features = { ...ALL_FEATURES_ENABLED }
  }
}

// 门禁已移除：除 DISABLED_FEATURES 里被显式关闭的模块外，一律返回 true。
// 保留函数签名，因为 App.vue / FloatingSphere.vue / Dashboard.vue /
// DashboardView.vue / Students.vue 等十余处模板仍在调用它。
export function hasFeature(featureName) {
  return !DISABLED_FEATURES.includes(featureName)
}

export async function activateLicense(licenseKey, selectedFeatures = null, contactInfo = {}) {
  const body = { license_key: licenseKey }
  if (selectedFeatures && selectedFeatures.length) {
    body.selected_features = selectedFeatures
  }
  if (contactInfo.organization_name) body.organization_name = contactInfo.organization_name
  if (contactInfo.contact_person) body.contact_person = contactInfo.contact_person
  if (contactInfo.contact_phone) body.contact_phone = contactInfo.contact_phone
  if (contactInfo.contact_email) body.contact_email = contactInfo.contact_email
  if (contactInfo.contact_wechat) body.contact_wechat = contactInfo.contact_wechat
  if (contactInfo.remarks) body.remarks = contactInfo.remarks
  const response = await api.post('/license/activate', body)
  await loadLicenseStatus()
  return response.data
}

export async function applyLicense(contactInfo, selectedFeatures, applyType = 'new', licenseType = '', referralCode = '') {
  const body = {
    organization_name: contactInfo.organization_name || '',
    contact_person: contactInfo.contact_person || '',
    contact_phone: contactInfo.contact_phone || '',
    contact_email: contactInfo.contact_email || '',
    contact_wechat: contactInfo.contact_wechat || '',
    remarks: contactInfo.remarks || '',
    selected_features: selectedFeatures || [],
    license_type: licenseType || '',
    apply_type: applyType,
    referral_code: referralCode || '',
  }
  const response = await api.post('/license/apply', body)
  return response.data
}

export async function previewAddonLicense(licenseKey) {
  const response = await api.post('/license/preview-addon', { license_key: licenseKey })
  return response.data
}

export async function deactivateLicense() {
  const response = await api.post('/license/deactivate')
  await loadLicenseStatus()
  return response.data
}

export async function deactivateFeature(featureName) {
  const response = await api.post(`/license/deactivate-feature/${featureName}`)
  await loadLicenseStatus()
  return response.data
}

export async function getMachineCode() {
  const response = await api.get('/license/machine-code')
  return response.data.machine_code
}

export async function notifySupplierView(data) {
  const response = await api.post('/license/notify-supplier-view', data)
  return response.data
}

export async function submitFeedback(data) {
  const response = await api.post('/license/feedback', data)
  return response.data
}

export async function requestReplaceLicense(data) {
  const response = await api.post('/license/request-replace', data)
  return response.data
}